<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Retrieve calculated GWAS colocalisation data — gwasColocalisation • otargen</title><!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png"><link rel="icon" type="”image/svg+xml”" href="../favicon.svg"><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" sizes="any" href="../favicon.ico"><link rel="manifest" href="../site.webmanifest"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Retrieve calculated GWAS colocalisation data — gwasColocalisation"><meta name="description" content="This function retrieves colocalisation data for a specific study locus from a GWAS study with other GWAS studies.
It returns a data frame of the studies that colocalise with the input study locus,
including details on the study, reported trait, index variant, and calculated colocalisation method outputs."><meta property="og:description" content="This function retrieves colocalisation data for a specific study locus from a GWAS study with other GWAS studies.
It returns a data frame of the studies that colocalise with the input study locus,
including details on the study, reported trait, index variant, and calculated colocalisation method outputs."><meta property="og:image" content="https://amirfeizi.github.io/otargen/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">otargen</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.0.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/web_only/Contributingguidelines.html">Contributing guidelines</a></li>
    <li><a class="dropdown-item" href="../articles/web_only/introduction.html">examples</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/amirfeizi/otargen/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Retrieve calculated GWAS colocalisation data</h1>
      <small class="dont-index">Source: <a href="https://github.com/amirfeizi/otargen/blob/main/R/GWASColocQuery.R" class="external-link"><code>R/GWASColocQuery.R</code></a></small>
      <div class="d-none name"><code>gwasColocalisation.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function retrieves colocalisation data for a specific study locus from a GWAS study with other GWAS studies.
It returns a data frame of the studies that colocalise with the input study locus,
including details on the study, reported trait, index variant, and calculated colocalisation method outputs.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">gwasColocalisation</span><span class="op">(</span><span class="va">study_locus_id</span>, size <span class="op">=</span> <span class="fl">500</span>, index <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-study-locus-id">study_locus_id<a class="anchor" aria-label="anchor" href="#arg-study-locus-id"></a></dt>
<dd><p>Character: Open Target Genetics generated ID for the study locus (e.g., "5a86bfd40d2ebecf6ce97bbe8a737512").</p></dd>


<dt id="arg-size">size<a class="anchor" aria-label="anchor" href="#arg-size"></a></dt>
<dd><p>Integer: Number of rows to fetch per page. Default: 500.</p></dd>


<dt id="arg-index">index<a class="anchor" aria-label="anchor" href="#arg-index"></a></dt>
<dd><p>Integer: Page index for pagination. Default: 0.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>Returns a data frame of the studies that colocalise with the input study locus. The table consists of the following data structure:</p><ul><li><p><code>study.studyId</code>: <em>Character vector</em>. Study identifier.</p></li>
<li><p><code>study.traitReported</code>: <em>Character vector</em>. Reported trait associated with the colocalisation.</p></li>
<li><p><code>study.projectId</code>: <em>Character vector</em>. Project identifier for the study.</p></li>
<li><p><code>study.publicationFirstAuthor</code>: <em>Character vector</em>. First author of the publication.</p></li>
<li><p><code>indexVariant.id</code>: <em>Character vector</em>. Index variant identifier.</p></li>
<li><p><code>indexVariant.position</code>: <em>Integer vector</em>. Index variant position.</p></li>
<li><p><code>indexVariant.chromosome</code>: <em>Character vector</em>. Index variant chromosome.</p></li>
<li><p><code>indexVariant.referenceAllele</code>: <em>Character vector</em>. Reference allele of the variant.</p></li>
<li><p><code>indexVariant.alternateAllele</code>: <em>Character vector</em>. Alternate allele of the variant.</p></li>
<li><p><code>pValueMantissa</code>: <em>Numeric vector</em>. Mantissa of the p-value for the colocalisation.</p></li>
<li><p><code>pValueExponent</code>: <em>Integer vector</em>. Exponent of the p-value for the colocalisation.</p></li>
<li><p><code>numberColocalisingVariants</code>: <em>Integer vector</em>. Number of colocalising variants.</p></li>
<li><p><code>colocalisationMethod</code>: <em>Character vector</em>. Method used for colocalisation analysis.</p></li>
<li><p><code>h3</code>: <em>Numeric vector</em>. H3 value associated with the colocalisation.</p></li>
<li><p><code>h4</code>: <em>Numeric vector</em>. H4 value associated with the colocalisation.</p></li>
<li><p><code>clpp</code>: <em>Numeric vector</em>. Colocalisation posterior probability.</p></li>
<li><p><code>betaRatioSignAverage</code>: <em>Numeric vector</em>. Average sign of the beta ratio.</p></li>
</ul></div>
    <div class="section level2">
    <h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a></h2>
    <p>Giambartolomei, Claudia et al. “Bayesian test for colocalisation between pairs of genetic association studies using summary statistics.” PLoS genetics vol. 10,5 e1004383. 15 May. 2014, doi:10.1371/journal.pgen.1004383</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span> <span class="co"># \dontrun{</span></span></span>
<span class="r-in"><span><span class="va">colocalisation_data</span> <span class="op">&lt;-</span> <span class="fu">gwasColocalisation</span><span class="op">(</span>study_locus_id <span class="op">=</span> <span class="st">"5a86bfd40d2ebecf6ce97bbe8a737512"</span>,</span></span>
<span class="r-in"><span> size <span class="op">=</span> <span class="fl">500</span>, index <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span> <span class="co"># }</span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Amir Feizi.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

